### PR TITLE
mgr/dashboard: Use pipe instead of calling function within template

### PR DESCRIPTION
**Pull request by votdev**
Refactor code. Use pipes instead of calling a function from within the template.

Fixes: https://tracker.ceph.com/issues/48051

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by epuertat — COMMENTED** [review]
Nice proposal! My question: rather than creating new pipes for every function, why not simply use a generic pipe for functions?:
* Either [ngx-pipe-function](https://www.npmjs.com/package/ngx-pipe-function)
* Or [this other PoC PR that I started](https://github.com/ceph/ceph/pull/36861) 

I'd rather prefer `ngx-pipe-function` over my PoC (168 weekly downloads vs. 0), even if it lacks support for multiple argument functions.

**Comment by votdev** [comment]
> Nice proposal! My question: rather than creating new pipes for every function, why not simply use a generic pipe for functions?:
> 
>     * Either [ngx-pipe-function](https://www.npmjs.com/package/ngx-pipe-function)
> 
>     * Or [this other PoC PR that I started](https://github.com/ceph/ceph/pull/36861)
> 
> 
> I'd rather prefer `ngx-pipe-function` over my PoC (168 weekly downloads vs. 0), even if it lacks support for multiple argument functions.

A collection of pipes looks also promising: https://fknop.gitbook.io/angular-pipes/

But i personally do not like to use external libraries if it is not really necessary. You have more control over your own code than when it comes from outside.

**Review by alfonsomthd — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/pipes/is-empty.pipe.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    expect(pipe).toBeTruthy();
+  });
+
+  it('should transform to true (1)', () => {
+    expect(pipe.transform(NaN)).toBeTruthy();
+  });
+
+  it('should transform to true (2)', () => {
+    expect(pipe.transform(undefined)).toBeTruthy();
+  });
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/pipes/is-empty.pipe.spec.ts:16]: What's the value of testing that the external `isEmpty()` function works as expected? [`lodash` already has more comprehensive unit tests](https://github.com/lodash/lodash/blob/master/test/isEmpty.js) for that purpose?

However, this is not testing the key thing this change is bringing in (reducing change detection events for functions in templates), which, for example, [`ngx-function-pipe` actually does](https://github.com/ArtemLanovyy/ngx-pipe-function/blob/master/__tests__/pipe-function.pipe.spec.ts#L73).

That's my point with using external libraries: you not only get 'working code' for free, you may also get higher quality (via more comprehensive unit tests), and a community supporting that and [reporting issues](https://github.com/ArtemLanovyy/ngx-pipe-function/issues).

**Comment by epuertat** [comment]
> But i personally do not like to use external libraries if it is not really necessary. You have more control over your own code than when it comes from outside.

IMHO software eng. is not about control, it's about effort vs. quality: we currently don't have this functionality implemented, so we don't need more control over something that we don't even have yet in our codebase. Instead, should we prefer having individual pipes, one by one, for each function we want to put in the HTML template?

> In this special case using a function pipe does not really help because this pipe is so generic that it might be used in more than one component, so calling a component function does not help, except you want to duplicate the called function (`isEmpty`) in several components.

Well, if you want to use another method you need to repeat this PR for every new method (5 files modified, tens of lines added...) vs `this._ = _`.

> I really like to see a foundation of pipes that can be used by the Dashboard developers.

Why only Ceph-Dashboard developers? We have a wider community out there.

**Comment by votdev** [comment]
> > But i personally do not like to use external libraries if it is not really necessary. You have more control over your own code than when it comes from outside.
> 
> IMHO software eng. is not about control, it's about effort vs. quality: we currently don't have this functionality implemented, so we don't need more control over something that we don't even have yet in our codebase. Instead, should we prefer having individual pipes, one by one, for each function we want to put in the HTML template?
> 
> > In this special case using a function pipe does not really help because this pipe is so generic that it might be used in more than one component, so calling a component function does not help, except you want to duplicate the called function (`isEmpty`) in several components.
> 
> Well, if you want to use another method you need to repeat this PR for every new method (5 files modified, tens of lines added...) vs `this._ = _`.
> 
> > I really like to see a foundation of pipes that can be used by the Dashboard developers.
> 
> Why only Ceph-Dashboard developers? We have a wider community out there.

I've opened https://github.com/ceph/ceph/pull/37993 that may replace this PR. The new PR is using `ngx-pipe-function` instead of creating a new `isEmpty` pipe.

**Comment by epuertat** [comment]
@votdev, thank you!

**Comment by votdev** [comment]
Closed in favor of https://github.com/ceph/ceph/pull/37993.

**Pull request closed without merging**